Project 3/match_frame.py

```diff
@@ -72,6 +72,47 @@ def show_ten_closest(frame_data, feature_summary, frame_num, description):
     cv2.destroyAllWindows()
     return
 
+def show_ten_quantized_closest(frame_data,frame_block_dict,target_frame_number, bins, description ):
+    target_frame = frame_data[target_frame_number-1]  #we need to create block histograms for this frame
+    target_frame_block_hist_dict = {}
+    top_ten_frames = [(0,100000000)] * 10 #sufficently large values that will be replace later
+
+    for block_x in range(0, len(target_frame), 8):  #create histogram for the given frame
+            for block_y in range(0, len(target_frame[block_x]), 8):
+                target_frame_block = target_frame[block_x:block_x+8, block_y:block_y+8]
+                target_frame_block_hist = cv2.calcHist([target_frame_block.astype(np.uint8)],[0],None,[bins],[0,256])
+                target_frame_block_hist_dict[block_x,block_y] = target_frame_block_hist
+
+    print("Comparing frames...")
+    for keyA in frame_block_dict:
+        if keyA == target_frame_number:  #dont compare the frame against itself
+            continue
+        else:
+            frame_score = 0
+            for keyB in frame_block_dict[keyA]:
+                block_hist = frame_block_dict[keyA][keyB[0],keyB[1]]
+                frame_score += cv2.compareHist(target_frame_block_hist_dict[keyB[0],keyB[1]],block_hist, 1)
+
+        largest_diff = 0
+        replace_index = -1
+        for i in range (0,10):
+            diff = top_ten_frames[i][1] - frame_score
+            if largest_diff < diff:
+                largest_diff = diff
+                replace_index = i
+        if replace_index >= 0:
+            top_ten_frames[replace_index] = keyA,largest_diff
+
+    top_ten_frames = list((x[0] for x in top_ten_frames))    #just need to the frame number, not the diff so we strip that out
+    for i in range(0,10):
+        # For each of those frame numbers, display the image in a window with the number
+        index = top_ten_frames[i]
+        rgb_target = cv2.cvtColor(frame_data[index].astype(np.uint8), cv2.COLOR_GRAY2BGR)
+        cv2.imshow(description + ' ' + str(i), rgb_target)
+        cv2.waitKey(0)
+
+    cv2.destroyAllWindows()
+
 
```

Comment:
This starts by building the block histograms for the target frame, then accesses the structures made above to compare the target frame histograms against all the frames histograms.